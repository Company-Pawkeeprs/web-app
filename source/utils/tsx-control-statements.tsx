import React from 'react'

type JsxChild =
    | string
    | boolean
    | number
    | null
    | undefined
    | JSX.Element
    | React.ReactNode
type JsxChildren = JsxChild | JsxChild[]

type PropsFor<T> = {
    items: T[]
    children: (item: T, index: number) => JsxChildren
}

export function For<T>({ children, items }: PropsFor<T>) {
    return (
        <React.Fragment>
            {items.map((item, index) => {
                return children(item, index)
            })}
        </React.Fragment>
    )
}

type PropsIf = {
    condition: boolean
    children: JsxChildren
}

export function If({ condition, children }: PropsIf) {
    if (!condition) return null

    return <React.Fragment>{children}</React.Fragment>
}

type PropsChoose = {
    value: any
    children: React.ReactNode
}

export function Choose({ value, children }: PropsChoose) {
    return (
        <React.Fragment>
            {React.Children.map(children, (child: any) => {
                if (child.props.value === value) {
                    return child.props.children
                }
            })}
        </React.Fragment>
    )
}

type PropsWhen = {
    value: any
    children: JsxChildren
}

export function When({ value, children }: PropsWhen) {
    return <React.Fragment>{children}</React.Fragment>
}
