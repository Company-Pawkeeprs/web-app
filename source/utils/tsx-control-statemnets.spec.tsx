import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Choose, For, If, When } from './tsx-control-statements';


describe('tsx-control-statements', () => {
    it('should render For (Unit)', () => {
        const items = [1, 2, 3]

        const { asFragment } = render(
            <For items={items}>
                {(item, index) => <div key={index}>{item}</div>}
            </For>,
        )

        expect(asFragment()).toMatchSnapshot()
    })

    it('should render If (Unit)', () => {
        const { asFragment } = render(
            <If condition={true}>
                <div>test</div>
            </If>,
        )

        expect(asFragment()).toMatchSnapshot()
    })

    it('should render Choose (Unit)', () => {
        const { asFragment } = render(
            <Choose value={1}>
                <When value={1}>
                    <div>test</div>
                </When>
            </Choose>,
        )

        expect(asFragment()).toMatchSnapshot()
    })

    it('should render null when condition is false (Unit)', async () => {
        render(
            <If condition={false}>
                <label htmlFor="username">Username</label><input id="username" type="text" />
            </If>,
        )


        const element = screen.queryByRole('textbox', { name: /test/i, hidden: true })
        expect(element).not.toBeInTheDocument()

    })

    // it('should render null when value is not matched (Unit)', () => {
    //     const { asFragment } = render(
    //         <Choose value={1}>
    //             <When value={2}>
    //                 <div>test</div>
    //             </When>
    //         </Choose>,
    //     )

    //     expect(asFragment()).toMatchSnapshot()
    //     expect(asFragment()).toBeNull()
    // })

})    
