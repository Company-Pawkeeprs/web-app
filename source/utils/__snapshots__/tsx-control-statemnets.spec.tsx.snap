// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tsx-control-statements > should render Choose (Unit) 1`] = `
<DocumentFragment>
  <div>
    test
  </div>
</DocumentFragment>
`;

exports[`tsx-control-statements > should render For (Unit) 1`] = `
<DocumentFragment>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</DocumentFragment>
`;

exports[`tsx-control-statements > should render If (Unit) 1`] = `
<DocumentFragment>
  <div>
    test
  </div>
</DocumentFragment>
`;
